Archive/Wiedijk100Theorems/SolutionOfCubicQuartic.lean

```diff
@@ -131,6 +157,104 @@ theorem cubic_eq_zero_iff_of_p_eq_zero (ha : a ≠ 0) (hω : IsPrimitiveRoot ω
+      have hw' : discrim 2 (-(2 * s)) (u + q / s) = w * w := by rw [discrim]; linear_combination -hw
+      rw [mul_eq_zero, quadratic_eq_zero_iff hi2 hv', quadratic_eq_zero_iff hi2 hw']
+      simp [(by norm_num : (2 : K) * 2 = 4), or_assoc, or_comm]
+
+/-- **The Solution of Quartic**.
+  The roots of a quartic polynomial when q is nonzero.
+  See Beyer W. H., CRC Standard Mathematical Tables and Formulae.
+  Here, u needs to satisfy the cubic resolvent. An explicit expression of u is possible using
+  the cubic formula, but would be too long. -/
+theorem quartic_eq_zero_iff (ha : a ≠ 0)
+    (hp : p = (8 * a * c - 3 * b ^ 2) / (8 * a ^ 2))
+    (hq : q = (b ^ 3 - 4 * a * b * c + 8 * a ^ 2 * d) / (8 * a ^ 3)) (hq_nonzero : q ≠ 0)
+    (hr : r =
+      (16 * a * b ^ 2 * c + 256 * a ^ 3 * e - 3 * b ^ 4 - 64 * a ^ 2 * b * d) / (256 * a ^ 4))
+    (hu : u ^ 3 - p * u ^ 2 - 4 * r * u + 4 * p * r - q ^ 2 = 0)
+    (hs : s ^ 2 = u - p)
+    (hv : v ^ 2 = 4 * s ^ 2 - 8 * (u - q / s))
+    (hw : w ^ 2 = 4 * s ^ 2 - 8 * (u + q / s)) (x : K) :
+    a * x ^ 4 + b * x ^ 3 + c * x ^ 2 + d * x + e = 0 ↔
+      x = (-2 * s - v) / 4 - b / (4 * a) ∨ x = (-2 * s + v) / 4 - b / (4 * a) ∨
+        x = (2 * s - w) / 4 - b / (4 * a) ∨ x = (2 * s + w) / 4 - b / (4 * a) := by
+  let y := x + b / (4 * a)
+  have h4 : (4 : K) = 2 ^ 2 := by norm_num
+  have h8 : (8 : K) = 2 ^ 3 := by norm_num
+  have h16 : (16 : K) = 2 ^ 4 := by norm_num
+  have h256 : (256 : K) = 2 ^ 8 := by norm_num
+  have h₁ : a * x ^ 4 + b * x ^ 3 + c * x ^ 2 + d * x + e =
+      a * (y ^ 4 + p * y ^ 2 + q * y + r) := by
+    rw [hp, hq, hr]
+    simp [field_simps, y, ha, h4, h8, h16, h256]; ring
+  have h₂ : ∀ x, a * x = 0 ↔ x = 0 := by intro x; simp [ha]
+  rw [h₁, h₂, quartic_depressed_eq_zero_iff hq_nonzero hu hs hv hw]
+  simp_rw [eq_sub_iff_add_eq]
+
+/-- The roots of a quartic polynomial when q equals zero. -/
+theorem quartic_eq_zero_iff_of_q_eq_zero (ha : a ≠ 0)
+    (hp : p = (8 * a * c - 3 * b ^ 2) / (8 * a ^ 2))
+    (hqz : b ^ 3 - 4 * a * b * c + 8 * a ^ 2 * d = 0)
+    (hr : r =
+      (16 * a * b ^ 2 * c + 256 * a ^ 3 * e - 3 * b ^ 4 - 64 * a ^ 2 * b * d) / (256 * a ^ 4))
+    (ht : t ^ 2 = p ^ 2 - 4 * r)
+    (hv : v ^ 2 = (-p + t) / 2)
+    (hw : w ^ 2 = (-p - t) / 2) (x : K) :
+    a * x ^ 4 + b * x ^ 3 + c * x ^ 2 + d * x + e = 0 ↔
+      x = v - b / (4 * a) ∨ x = -v - b / (4 * a) ∨ x = w - b / (4 * a) ∨ x = -w - b / (4 * a) := by
+  let y := x + b / (4 * a)
+  have h4 : (4 : K) = 2 ^ 2 := by norm_num
+  have h8 : (8 : K) = 2 ^ 3 := by norm_num
+  have h16 : (16 : K) = 2 ^ 4 := by norm_num
+  have h256 : (256 : K) = 2 ^ 8 := by norm_num
+  have h₁ : a * x ^ 4 + b * x ^ 3 + c * x ^ 2 + d * x + e = a * (y ^ 4 + p * y ^ 2 + r) := by
+    rw [hp, hr]
+    simp [field_simps, y, ha, h4, h8, h16, h256]
+    linear_combination (1048576 * a ^ 10 * x + 262144 * a ^ 9 * b) * hqz
+  have h₂ : ∀ x, a * x = 0 ↔ x = 0 := by intro x; simp [ha]
```

Comment:
This is a lemma that you should be able to find and use directly, I think; also above